### PR TITLE
CRM-20311 - Membership cancelation via contribution doesn't create me…

### PR DESCRIPTION
**Pull request by sdragnev**
…mbership status change activity

Proposed fix adds the activity bit of code to the deprecated function in CRM_Contribute_BAO_Contribution.
This isn't a great fix but given that the function is deprecated anyway, anything long term will depend on how this whole transition is handled in the future.

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
test this please

File: CRM/Contribute/BAO/Contribution.php [hunk 1]
```diff
@@ -1722,8 +1722,24 @@ public static function transitionComponents($params, $processContributionObject
...
+            $newStatus = array_search('Cancelled', $membershipStatuses);
+
+            // Create activity
+            $allStatus = CRM_Member_BAO_Membership::buildOptions('status_id', 'get');
+            $activityParam = array(
+              'subject' => "Status changed from {$allStatus[$membership->status_id]} to {$allStatus[$newStatus]}",
+              'source_contact_id' => CRM_Core_Session::singleton()->get('userID'),
+              'target_contact_id' => $membership->contact_id,
+              'source_record_id' => $membership->id,
+              'activity_type_id' => CRM_Core_PseudoConstant::getKey('CRM_Activity_BAO_Activity', 'activity_type_id', 'Change Membership Status'),
```
Comment (eileenmcnaughton) [inline CRM/Contribute/BAO/Contribution.php]: for these rows you don't need to resolve the values ie

'activity_type_id' => 'Change Membership Status',
'status_id' => 'Completed',
'priority_id' => 'Normal',
'activity_date_time' =>'now', 

some of those might be the defaults & not need setting.

The api handles those & they are more readable and it removes the hard-coding.

Note I have only looked at this from a code POV - not checked there is agreement on the problem.

What I would REALLY like to do is help you to write a unit test for this.

Comment (sdragnev) [inline CRM/Contribute/BAO/Contribution.php]: Yeah, good call. Removed the look up and added labels instead of hardcoded IDs.

**Comment by seamuslee001** [comment]
Test this please

**Comment by petednz** [comment]
not sure how i 'assign' - happy for someone to assign to jitendra to test

**Comment by eileenmcnaughton** [comment]
@petednz You can comment on a PR @jitendrapurohit  "please write a unit test for this & confirm it is good to merge" - then people will see you have asked him to do it.

**Comment by petednz** [comment]
@jitendrapurohit "please write a unit test for this & confirm it is good to merge"

**Comment by jitendrapurohit** [comment]
Added test here - https://github.com/civicrm/civicrm-core/pull/10054. There are still some issues related to membership contribution, but that seems unrelated to this PR. Hence, this could be merged, will create another ticket to address the others.

Thanks.

**Review by jitendrapurohit — APPROVED** [review]

**Comment by eileenmcnaughton** [comment]
Thanks @jitendrapurohit, I would also add that I looked at the code & felt it was on the mark, although Stan made a couple of changes I suggested - so agree all good to merge!

**Comment by jitendrapurohit** [comment]
raised https://issues.civicrm.org/jira/browse/CRM-20342 and https://issues.civicrm.org/jira/browse/CRM-20343 for related issues.

**Comment by eileenmcnaughton** [comment]
@jitendrapurohit the first of those looks pretty serious! Are you able to work on one or both of them?

**Pull request merged**